internal/dataplane/translator/translate_ingress.go

```diff
@@ -81,53 +84,133 @@ func (t *Translator) ingressRulesFromIngressV1() ingressRules {
-				ReadTimeout:    kong.Int(DefaultServiceTimeout),
-				WriteTimeout:   kong.Int(DefaultServiceTimeout),
-				Retries:        kong.Int(DefaultRetries),
-				Tags:           util.GenerateTagsForObject(&ingress),
-			},
-			Namespace: ingress.Namespace,
-			Backends: []kongstate.ServiceBackend{{
-				Name:    defaultBackend.Service.Name,
-				PortDef: port,
-			}},
-			Parent: &ingress,
+		defaultBackend := ingress.Spec.DefaultBackend
+		route := translateIngressDefaultBackendRoute(&ingress, util.GenerateTagsForObject(&ingress), features.ExpressionRoutes)
+
+		// If the default backend is defined as an arbitrary resource, then we need handle it differently.
+		if resource := defaultBackend.Resource; resource != nil {
+			return translateIngressDefaultBackendResource(
+				resource,
+				ingress,
+				route,
+				storer,
+				failuresCollector,
+				features,
+			)
 		}
-		r := translateIngressDefaultBackendRoute(&ingress, util.GenerateTagsForObject(&ingress), expressionRoutes)
-		service.Routes = append(service.Routes, *r)
-		return service, true
+
+		// Otherwise, the default backend is defined as a Kubernetes Service.
+		return translateIngressDefaultBackendService(ingress, route)
 	}
 
 	return kongstate.Service{}, false
 }
 
+func translateIngressDefaultBackendResource(
+	resource *corev1.TypedLocalObjectReference,
+	ingress netv1.Ingress,
+	route *kongstate.Route,
+	storer store.Storer,
+	failuresCollector *failures.ResourceFailuresCollector,
+	features FeatureFlags,
+) (kongstate.Service, bool) {
+	if !subtranslator.IsKongServiceFacade(resource) {
+		gk := resource.Kind
+		if resource.APIGroup != nil {
+			gk = *resource.APIGroup + "/" + gk
+		}
+		failuresCollector.PushResourceFailure(fmt.Sprintf("default backend: unknown resource type %s", gk), &ingress)
+		return kongstate.Service{}, false
+	}
+	if !features.KongServiceFacade {
+		failuresCollector.PushResourceFailure(
+			fmt.Sprintf("default backend: KongServiceFacade is not enabled, please set the %q feature gate to 'true' to enable it", featuregates.KongServiceFacade),
+			&ingress,
+		)
+		return kongstate.Service{}, false
+	}
+	facade, err := storer.GetKongServiceFacade(ingress.Namespace, resource.Name)
+	if err != nil {
+		failuresCollector.PushResourceFailure(
+			fmt.Sprintf("default backend: KongServiceFacade %q could not be fetched: %s", resource.Name, err),
+			&ingress,
+		)
+		return kongstate.Service{}, false
+	}
+
+	serviceName := fmt.Sprintf("%s.%s.svc.facade", ingress.Namespace, resource.Name)
+	return kongstate.Service{
+		Service: kong.Service{
+			Name:           kong.String(serviceName),
+			Host:           kong.String(serviceName),
+			Port:           kong.Int(DefaultHTTPPort),
+			Protocol:       kong.String("http"),
+			ConnectTimeout: kong.Int(DefaultServiceTimeout),
+			ReadTimeout:    kong.Int(DefaultServiceTimeout),
+			WriteTimeout:   kong.Int(DefaultServiceTimeout),
+			Retries:        kong.Int(DefaultRetries),
```

Comment:
I think we should add tags to the generated service to tell the Kong gateway that this is generated from some k8s resources (is the tags from the `KongServiceFacade` a proper value?)

Reply:
Kong Service tags are populated later in the `Translator` pipeline - namely in `ingressRules.populateServices -> `ingressRules.generateKongServiceTags`.

To make it clear, I added comments explaining that and covered generated tags in `Translator`'s unit tests (both for `Service` and `KongServiceFacade` backends).